### DISCUSSION TITLE
How to add additional avisynth filter

### DISCUSSION THREAD

**Discussion by Cyberavater** [staxrip/staxrip#741] [Q&A] [answered]
Hi, I want to add ExactDedup http://avisynth.nl/index.php/ExactDedup, how do I do that?

**Comment by JJKylee** [comment, marked as answer]
Refer to [this wiki](https://github.com/staxrip/staxrip/wiki/How-to-register-DGTonemap-&-DGHDRtoSDR-filters-in-StaxRip) to get a rough idea on working with a **_binary_** filter plugin.

1. Put `dll`s in a subfolder under `%settings_dir%Plugins\AviSynth`.
2. Use `LoadPlugin` call in the AVS filter profile referring to a working script code example. You may need to use StaxRip macros in composing a profile.

For a **_script_** filter plugin that's provided as an `avsi` script file, use `Import` in the above Step 2.

---

For `ExactDedup`, we have a working script code example [here](https://forum.doom9.org/showthread.php?p=1944292#post1944292) on Doom9. But in order to apply the filter properly, it seems that we need to get [`RT_Stats`](http://avisynth.nl/index.php/RT_Stats) as well.

So first get the most recent `RT_Stats` and `ExactDedup` plugins from [StainlessS's MediaFire archive](https://www.mediafire.com/folder/hb26mthbjz7z6/StainlessS). (`RT_Stats_25&26_x86_x64_dll_v2.00Beta13_20201229.zip` and `ExactDedup_25_26_x86_x64_dll_v0-07Beta_20210604.zip`)

Unzip and put `ExactDeDup_x64.dll` and `RT_Stats_x64.dll` under `%settings_dir%Plugins\AviSynth\ExactDeDup\` and `%settings_dir%Plugins\AviSynth\RT_Stats\` respectively.

Compose a filter profile for `ExactDedup` as follows referring to the working script example by Hotte in Doom9:

```

DeDup | ExactDedup =
    LoadPlugin("%settings_dir%Plugins\AviSynth\RT_Stats\RT_Stats_x64.dll")
    LoadPlugin("%settings_dir%Plugins\AviSynth\ExactDeDup\ExactDeDup_x64.dll")
    DUPTXT = "%temp_file%_dupinfo_dd.txt"
    DUPTXT = RT_GetFullPathName(DUPTXT)
    TIMETXT = "%temp_file%_timecodes_dd.txt"
    TIMETXT = RT_GetFullPathName(TIMETXT)
    ExactDedup(firstpass=true, dupinfo=DUPTXT, times=TIMETXT, maxdupcount=1000000, keeplastframe=false)
    RT_ForceProcess()
    ExactDedup(firstpass=false, dupinfo=DUPTXT, times=TIMETXT, maxdupcount=1000000, keeplastframe=false)
    # Don't forget to mux TIMETXT timecodes file in the output mkv

```

Put it under, eg, **_[Restoration]_** category in _AVS Filter Profiles_. (Blank lines above and below)

Apply `ExactDedup` in the filter menu, and _make sure to mux the timecodes file generated by the filter_(`%temp_file%_timecodes_dd.txt`) in `mkvmerge`.

**Reply by Cyberavater** [reply]
Thanks for your wonderful reply, I really appreciate it! On a side note, which would be better? ExactDedup or Dedup?

**Reply by Cyberavater** [reply]
I'm having this issue,

```
---------------------------- Muxing to MKV ----------------------------

mkvmerge 56.1

D:\Softwares\StaxRip-v2.6.0-x64\Apps\Support\MKVToolNix\mkvmerge.exe -o D:\test\Videos\Ajmeri\AVSEQ01_new.mkv --timestamps 0:"%temp_file%_timecodes_dd.txt" D:\test\Videos\Ajmeri\AVSEQ01_temp\AVSEQ01_new_out.h264 --audio-tracks 0 --language 0:eng --sync 0:-80 --default-track 0:0 --forced-track 0:0 "D:\test\Videos\Ajmeri\AVSEQ01_temp\ID1 -80ms.aac" --ui-language en

mkvmerge v56.1.0 ('My Friend') 64-bit
'D:\test\Videos\Ajmeri\AVSEQ01_temp\AVSEQ01_new_out.h264': Using the demultiplexer for the format 'AVC/H.264'.
'D:\test\Videos\Ajmeri\AVSEQ01_temp\ID1 -80ms.aac': Using the demultiplexer for the format 'AAC'.
Error: The timestamp file '%temp_file%_timecodes_dd.txt' could not be opened for reading.

Start:    7:55:12 PM
End:      7:55:13 PM
Duration: 00:00:00
```

> Apply ExactDedup in the filter menu, and make sure to mux the timecodes file generated by the filter(%temp_file%_timecodes_dd.txt) in mkvmerge.

What I've done, Container Option > Option > Timestamp> %temp_file%_timecodes_dd.txt

**Reply by JJKylee** [reply]
> What I've done, Container Option > Option > Timestamp> %temp_file%_timecodes_dd.txt

No. `%temp_file%` is just a StaxRip _**macro**_. You _**shouldn't**_ put it _**as is**_ in the path in _Container Options_ since StaxRip does not expand it like in filter profiles, etc. I just used it as a _notation_ for the actual timecodes file.

Browse for `<INPUT_FILE_NAME>_timecodes_dd.txt` in the `<Temp>` folder of the project in _Container Options_ > _Options_ > _Timestamps_ of `mkvmerge` and select it.

> On a side note, which would be better? ExactDedup or Dedup?

I don't know. Maybe it depends on the source and other things. Try them out yourself.

**Reply by Cyberavater** [reply]
> Browse for <INPUT_FILE_NAME>_timecodes_dd.txt in the <Temp> folder of the project in Container Options > Options > Timestamps of mkvmerge and select it.

So, it'll only work per file I assume? not when I'm doing it in batch?

**Reply by JJKylee** [reply]
Unfortunately, no.

If you want StaxRip to expand macros in _Container Options_ as well as in _Filter Profiles_, etc., you need to submit a **feature request** in the Issue Tracker. But of course whether it will be accepted or not depends on @Dendraspis's decision, though. 😓

**Reply by JJKylee** [reply]
`ExactDeDup` will be included in the next release (v2.7.0).

But `DeDup` will not, because its structure is too old to be profiled in just one script run.

Refer to the [manual in AviSynth wiki](http://avisynth.nl/index.php/DeDup), and run two different `avs` scripts in a row (2 passes) to get the results by `DeDup`.

**Comment by JJKylee** [comment]
For reference, another _script_ filter `FillDropsAlt` is available for fixing dup frames.

johnmeyer's original version: https://forum.doom9.org/showthread.php?p=1775184#post1775184

[Here](https://forum.videohelp.com/threads/402416-repair-Videos-with-duplicates-and-dropped-frames#post2624941) on VideoHelp, I explained how to apply (slightly modified) Selur's mod in StaxRip.

I'm not sure how effective this filter is though, so don't ask me which one to apply. 😓

---

[UPDATE]

Renamed `FillDropsAlt` back to `FillDrops` (for maintenance), and ported it to VapourSynth as well.
`FillDrops` for both AviSynth and VapourSynth will be included in the next release (v2.7.0).
